### PR TITLE
Fix spack versions

### LINKED ISSUE #3653: spack versions broken
I just ran `spack versions` on every single package in Spack and not a single one of them found a remote version. I've tested this on multiple machines with different versions of Python. Did something break in develop?

### PR DESCRIPTION
**Pull request by adamjstewart**
Fixes #3653.

### CODE REVIEW

**Comment by adamjstewart** [comment]
@tgamblin What's the policy on unit tests that require internet access? Also, `spack versions` is difficult to unit test because we blindly except all Errors. That should probably change. There's also the problem of what happens when the website goes down? We have to pick some very reliable, very important package to test.

**Comment by adamjstewart** [comment]
Oh yeah, still working on getting `spack versions` to work in Python 3 btw:
```
$ python3 bin/spack --debug versions nco
==> '/usr/local/bin/git' 'ls-tree' '--name-only' 'HEAD'
==> Error in _spider: Can't convert 'bytes' object to str implicitly
==> Safe versions (already checksummed):
  4.6.5  4.6.4  4.6.3  4.6.2  4.6.1  4.5.5
==> Remote versions (not yet checksummed):
  Found no versions for nco
==> Check the list_url and list_depth attribute on the package to help Spack find versions.
```

**Comment by adamjstewart** [comment]
Found the fix: http://stackoverflow.com/a/16699532/5828163

File: lib/spack/spack/util/web.py [hunk 1]
```diff
@@ -38,7 +39,7 @@
     from html.parser import HTMLParser
 
     # Also, HTMLParseError is deprecated and never raised.
-    class HTMLParseError:
+    class HTMLParseError(BaseException):
```
Comment (alalazo) [inline lib/spack/spack/util/web.py]: Minor thing: I think it's better to inherit from `Exception`, `BaseException` is usually not meant for user defined classes. See [here](https://docs.python.org/2/library/exceptions.html).

**Comment by alalazo** [comment]
@adamjstewart Not sure if this makes sense, but what if we create a URL within `tests/data` that starts with `file://`? Would this exercise this part of the code without requiring internet access?

**Comment by adamjstewart** [comment]
@alalazo I think I had problems when trying to run `spack versions` on a local `file://` URL. See #2898.

**Comment by adamjstewart** [comment]
Unrelated, but I just noticed that we're on the official Python packaging website:
https://packaging.python.org/key_projects/#spack

**Comment by citibeth** [comment]
Should we be using `py-six` to provide Python 2/3 compatibility?  It's a single source file that can be easily included in the Spack distro.

**Comment by adamjstewart** [comment]
@citibeth We are.

**Comment by tgamblin** [comment]
> Unrelated, but I just noticed that we're on the official Python packaging website:
> https://packaging.python.org/key_projects/#spack

Is it as awesome if I did it?
https://github.com/pypa/python-packaging-user-guide/pull/252

We get ~10 hits/2wks from that 😄

**Comment by adamjstewart** [comment]
Next step, release on PyPI.

**Comment by tgamblin** [comment]
There is a guy who already has the spack name -- he let me have the readthedocs site but he still has the PyPI name :(

**Comment by citibeth** [comment]
How will releasing on PyPI improve the usability or availability of Spack?

**Comment by adamjstewart** [comment]
> There is a guy who already has the spack name

Ha, his [Sphere PACKing](https://github.com/wackywendell/spack) repo has a single star. That's frustrating.

**Comment by citibeth** [comment]
Actually, it looks like just the kind of software we would want to package
and include in our Spack.  Any way we could change our name?

Maybe we could become "Spackle."  Unfortunately, that is already taken too,
but a package manager no less!  (Although it's been dead for 5 years).

https://github.com/kristi/spackle

Or maybe we just change our GitHub-name, while keeping our main name as
"Spack".  For example, could we be `spackinst` in GitHub, PyPI, etc?

**Comment by tgamblin** [comment]
I think I will try to negotiate with the other spack author again.  I feel bad asking him for the name, but hey, we have some branding going on here.

**Comment by tgamblin** [comment]
@alalazo @adamjstewart: I pushed tests for the web crawling and archive discovery.  This turned up some bugs in `_spider`, as well as in how `list_depth` was being used.  It was defined in the docs as the number of links followed, but it was implemented as that minus one.  I fixed that as well.

Hoping this increases the coverage a bit too.

**Comment by tgamblin** [comment]
@adamjstewart: before you merge this one, let's rebase and make it 2 commits.

**Pull request merged**